src/lib/zexe_backend/zexe_backend_common/plonk_constraint_system.ml

```diff
@@ -0,0 +1,379 @@
+          |> Sequence.fold ~init:(c0, i0, [], 0)
+                ~f:(fun (acc, i, ts, n) (c, j) ->
+                  if Int.equal i j then (Fp.add acc c, i, ts, n)
+                  else (c, j, (acc, i) :: ts, n + 1) )
+        in
+        List.rev ((acc, i) :: ts)
+      in
+      match terms with
+      | [] -> assert false
+      | [(x, i)] -> (x, `Var (V.External i))
+      | (s1, x1) :: tl ->
+        let (s2, x2) = completely_reduce sys tl in
+        let res = V.create_internal () in
+        add_generic_constraint sys ?constant
+          (s1, External x1) (s2, x2) ~output:(neg_one, res) ;
+        (Fp.one, `Var res)
+  ;;
+
+  let add_constraint ?label:_ sys
+      (constr : (Fp.t Snarky_backendless.Cvar.t, Fp.t) Snarky_backendless.Constraint.basic) =
+
+    let red = reduce_lincom sys in
+    match constr with
+
+    | Snarky_backendless.Constraint.Boolean x ->
+      let (s, x) = red x in
+      ( 
+        match x with
+        | `Var x -> add_generic_constraint sys ~mul:Fp.(negate (square s)) (s, x) (s, x)
+        | `Constant ->  ()
+      )
+
+    | Snarky_backendless.Constraint.Equal (x, y) ->
+      let (sx, xx), (sy, xy) = red x, red y in
+      ( 
+        match xx, xy with
+        | `Var xx, `Var xy -> add_generic_constraint sys (Fp.negate sx, xx) (sy, xy)
+        | _, _ ->  ()
+      )
+
+    | Snarky_backendless.Constraint.Square (x, y) ->
+      let (sx, xx), (sy, xy) = red x, red y in
+      ( 
+        match xx, xy with
+        | `Var xx, `Var xy -> add_generic_constraint sys ~mul:Fp.(sx * sx) ~output:(Fp.negate sy, xy) (sx, xx) (sx, xx)
+        | _, _ ->  ()
+      )
+
+    | Snarky_backendless.Constraint.R1CS (x, y, z) ->
+      let (sx, xx), (sy, xy), (sz, xz) = red x, red y, red z in
+      ( 
+        match xx, xy, xz with
+        | `Var xx, `Var xy, `Var xz -> add_generic_constraint sys ~mul:Fp.(sx * sy) ~output:(Fp.negate sz, xz) (sx, xx) (sy, xy)
+        | _, _, _ ->  ()
+      )
+
+    | Plonk_constraint.T (Poseidon { start; state }) ->
+      let add_round_state array = Array.iteri ~f:
+          (
+            fun i x -> match snd (red x) with
+              | `Var x -> V.Table.add_multi sys.equivalence_classes ~key:x ~data:{ row= sys.next_row; col= i } ;
+              | _ ->  ()
+          ) array;
+        sys.next_row <- sys.next_row + 1
+      in
+      add_round_state start;
+      Array.iter ~f:
+        (
+          fun state -> 
+            (* Backend.add_poseidon_constraint sys.backend sys.next_row *)
+            add_round_state state
+        ) state;
+      ()
+
+    | Plonk_constraint.T (EC_add { p1; p2; p3 }) ->
+      Array.iteri ~f:
+        (
+          fun i (x, y) -> match snd (red x), snd (red y) with
+            | `Var x, `Var y ->
+              V.Table.add_multi sys.equivalence_classes ~key:y ~data:{ row= sys.next_row; col= i } ;
+              V.Table.add_multi sys.equivalence_classes ~key:x ~data:{ row= sys.next_row + 1; col= i } ;
+            | _, _ ->  ()
```

Comment:
Need to handle the constant case (if only trivially by allocating new internal vars and adding generic constraints asserting they are equal to a constant)